src/ir/effects.h

```diff
@@ -422,14 +424,25 @@ class EffectAnalyzer {
         return;
       }
 
+      if (curr->isReturn) {
+        parent.branchesOut = true;
+      }
+
+      if (parent.funcEffectsMap) {
+        auto iter = parent.funcEffectsMap->find(curr->target);
+        if (iter != parent.funcEffectsMap->end()) {
+          // We have effect information for this call target, and can just use
+          // that.
+          parent.mergeIn(iter->second);
+          return;
+        }
+      }
+
       parent.calls = true;
       // When EH is enabled, any call can throw.
       if (parent.features.hasExceptionHandling() && parent.tryDepth == 0) {
```

Comment:
Hmm, actually maybe we can use `tryDepth` even with the global effect info...

Reply:
Can you elaborate?

Reply:
I was thinking that if `tryDepth > 0` then even if the call's effects say "I can throw", we could ignore that (since it will be caught). I added a testcase for that now, see https://github.com/WebAssembly/binaryen/pull/5040/commits/a173e67ae802ea9d8cf32582ab86dfada7f006f0 (and the commit before it adds the optimization.

Reply:
Yeah I think this is handled already, because we don't even set `throws_` in case it's within a `try`-`catch`.